server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -20,19 +28,57 @@ public class RemoteStoreRefreshListener implements ReferenceManager.RefreshListe
+    private final Set<String> filesUploadedToRemoteStore;
+    private static final Logger logger = LogManager.getLogger(RemoteStoreRefreshListener.class);
 
-    public RemoteStoreRefreshListener(Directory storeDirectory, Directory remoteDirectory) {
+    public RemoteStoreRefreshListener(Directory storeDirectory, Directory remoteDirectory) throws IOException {
         this.storeDirectory = storeDirectory;
         this.remoteDirectory = remoteDirectory;
+        // ToDo: Handle failures in reading list of files (GitHub #3397)
+        this.filesUploadedToRemoteStore = new HashSet<>(Arrays.asList(remoteDirectory.listAll()));
     }
 
     @Override
     public void beforeRefresh() throws IOException {
-        // ToDo Add implementation
+        // Do Nothing
     }
 
+    /**
+     * Upload new segment files created as part of the last refresh to the remote segment store.
+     * The method also deletes segment files from remote store which are not part of local filesystem.
+     * @param didRefresh true if the refresh opened a new reference
+     * @throws IOException in case of I/O error in reading list of local files
+     */
     @Override
     public void afterRefresh(boolean didRefresh) throws IOException {
-        // ToDo Add implementation
+        if (didRefresh) {
+            Set<String> localFiles = Arrays.stream(storeDirectory.listAll()).collect(Collectors.toSet());
+            localFiles.stream().filter(file -> !filesUploadedToRemoteStore.contains(file)).forEach(file -> {
+                try {
+                    remoteDirectory.copyFrom(storeDirectory, file, file, IOContext.DEFAULT);
+                    filesUploadedToRemoteStore.add(file);
+                } catch (NoSuchFileException e) {
+                    logger.info("The file {} does not exist anymore. It can happen in case of temp files", file);
+                } catch (IOException e) {
+                    // ToDO: Handle transient and permanent un-availability of the remote store (GitHub #3397)
+                    logger.warn("Exception while uploading file {} to the remote segment store", file);
+                }
+            });
+
+            Set<String> remoteFilesToBeDeleted = new HashSet<>();
+            // ToDo: Instead of deleting files in sync, mark them and delete in async/periodic flow (GitHub #3142)
+            filesUploadedToRemoteStore.stream().filter(file -> !localFiles.contains(file)).forEach(file -> {
+                try {
+                    remoteDirectory.deleteFile(file);
+                    remoteFilesToBeDeleted.add(file);
+                } catch (IOException e) {
+                    // ToDO: Handle transient and permanent un-availability of the remote store (GitHub #3397)
+                    logger.warn("Exception while deleting file {} from the remote segment store", file);
+                }
+            });
+
+            remoteFilesToBeDeleted.forEach(filesUploadedToRemoteStore::remove);
+        }
```

Comment:
1. Do we need to ensure this happens within a lock
2. let log exception as well
3. Can we do batch deletions outside of upload flow without blocking subsequent uploads?

Reply:
> Do we need to ensure this happens within a lock

`ReferenceManager.doMaybeRefresh()` acquires lock before notifying refresh listeners

> let log exception as well

Will add exception. Most probably it will change based on the exception handling.

> Can we do batch deletions outside of upload flow without blocking subsequent uploads?

Yes, that is the plan. Added as ToDo before the delete block. https://github.com/opensearch-project/OpenSearch/issues/3142